component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/validator/JWTValidator.java

```diff
@@ -72,185 +73,325 @@
+            }
 
-        if (isNotEmpty(validIssuer) && !validIssuer.equals(jwtIssuer)) {
-            return logAndReturnFalse("Invalid Issuer:" + jwtIssuer + " in the given JSON Web Token.");
-        }
+            String jwtIssuer = claimsSet.getIssuer();
+            String jwtSubject = resolveSubject(claimsSet);
+            List<String> audience = claimsSet.getAudience();
+            Date expirationTime = claimsSet.getExpirationTime();
+            String jti = claimsSet.getJWTID();
+            Date nbf = claimsSet.getNotBeforeTime();
+            Date issuedAtTime = claimsSet.getIssueTime();
+            long currentTimeInMillis = System.currentTimeMillis();
+            long timeStampSkewMillis = OAuthServerConfiguration.getInstance().getTimeStampSkewInSeconds() * 1000;
+            OAuthAppDO oAuthAppDO = getOAuthAppDO(jwtSubject);
+            String consumerKey = oAuthAppDO.getOauthConsumerKey();
+            String tenantDomain = oAuthAppDO.getUser().getTenantDomain();
+            if (!validateMandatoryFeilds(mandatoryClaims, claimsSet)) {
+                return false;
+            }
 
-        if (!validateJTI(signedJWT, jti, currentTimeInMillis, timeStampSkewMillis, expirationTime.getTime(),
-                issuedAtTime.getTime())) {
-            return false;
-        }
+            //Validate issuer and subject.
+            if (!validateIssuer(jwtIssuer, consumerKey) || !validateSubject(jwtSubject, consumerKey)) {
+                if (log.isDebugEnabled()) {
+                    log.debug("Invalid jwtIssuer: " + jwtIssuer + " or jwtSubject: " + jwtSubject + " is found in the " +
+                            "assertion." + "Expected value for issuer and subject is: " + oAuthAppDO.getOauthConsumerKey());
+                }
+                return false;
+            }
 
-        //validate whether the subject is client_id
-        OAuthAppDO oAuthAppDO = getOAuthAppDO(jwtSubject);
-        if (oAuthAppDO == null) {
-            return logAndReturnFalse("Unable to find OAuth application with provided JWT information with subject:"
-                    + jwtSubject);
+            // Get audience.
+            String validAud = getValidAudience(tenantDomain);
+            long expTime = 0;
+            long issuedTime = 0;
+            if (expirationTime != null) {
+                expTime = expirationTime.getTime();
+            }
+            if (issuedAtTime != null) {
+                issuedTime = issuedAtTime.getTime();
+            }
+
+            //Validate signature validation, audience, nbf,exp time, jti.
+            if (!validateAudience(validAud, audience) ||
+                    !validateJWTWithExpTime(expirationTime, currentTimeInMillis, timeStampSkewMillis) ||
+                    !validateNotBeforeClaim(currentTimeInMillis, timeStampSkewMillis, nbf) ||
+                    !validateJTI(signedJWT, jti, currentTimeInMillis, timeStampSkewMillis, expTime, issuedTime) ||
```

Comment:
validateJTI is to prevent token reuse, right?
We can move it up (after validateIssuer) , so that if token is played before no need to do other validations.